File: csharp/clock/Clock.cs
using System;
using System.Text;

public class Clock
{
    private readonly int _minutes;
    private const int HoursPerDay = 24;
    private const int MinutesPerHour = 60;

    public Clock(int hours, int minutes)
    {
        _minutes = minutes + (hours * MinutesPerHour);

        while (_minutes < 0)
        {
            _minutes += MinutesPerHour * HoursPerDay;
        }
    }

    public Clock Add(int minutesToAdd) => new Clock(0, _minutes + minutesToAdd);

    public Clock Subtract(int minutesToSubtract) => Add(0 - minutesToSubtract);

    public override string ToString() => $"{Hours():00}:{Minutes():00}";

    public override bool Equals(object obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        return obj.GetType() == this.GetType() && Equals((Clock) obj);
    }

    public override int GetHashCode()
    {
        unchecked
        {
            return (Hours() * 397) ^ Minutes();
        }
    }

    private bool Equals(Clock other) => Hours() == other.Hours() && Minutes() == other.Minutes();

    private int Minutes() => _minutes % MinutesPerHour;

    private int Hours() => (_minutes / MinutesPerHour) % HoursPerDay;

}

File: java/clock/src/main/java/Clock.java
class Clock {

    private int totalMinutes;

    public Clock(int hours, int minutes) {
        this.add(hours * 60 + minutes);
    }

    public void add(int minutes) {
        totalMinutes += minutes;
    }

    @Override
    public String toString() {
        return String.format("%02d:%02d", getHours(), getMinutes());
    }

    private int getMinutes() {
        return Math.floorMod(totalMinutes, 60);
    }

    private int getHours() {
        return Math.floorMod(Math.floorDiv(totalMinutes,  60), 24);
    }

    @Override
    public boolean equals(Object other) {
        // No tests for checking if obj is null, not a Clock, or this same Clock
        return this.hashCode() == other.hashCode();
    }

    @Override
    public int hashCode() {
        return toString().hashCode();
    }
}